pkg/telemetrylogs/json_query_builder.go

```diff
@@ -0,0 +1,660 @@
+package telemetrylogs
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"slices"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	schemamigrator "github.com/SigNoz/signoz-otel-collector/cmd/signozschemamigrator/schema_migrator"
+	"github.com/SigNoz/signoz-otel-collector/utils"
+	"github.com/SigNoz/signoz/pkg/errors"
+	"github.com/SigNoz/signoz/pkg/telemetrystore"
+	qbtypes "github.com/SigNoz/signoz/pkg/types/querybuildertypes/querybuildertypesv5"
+	"github.com/SigNoz/signoz/pkg/types/telemetrytypes"
+	"github.com/huandu/go-sqlbuilder"
+)
+
+var (
+	CodeCurrentNodeNil           = errors.MustNewCode("current_node_nil")
+	CodeNextNodeNil              = errors.MustNewCode("next_node_nil")
+	CodeNestedExpressionsEmpty   = errors.MustNewCode("nested_expressions_empty")
+	CodeGroupByPlanEmpty         = errors.MustNewCode("group_by_plan_empty")
+	CodeArrayMapExpressionsEmpty = errors.MustNewCode("array_map_expressions_empty")
+	CodePromotedPlanMissing      = errors.MustNewCode("promoted_plan_missing")
+)
+
+// Operator intent helpers
+func isMembershipContains(op qbtypes.FilterOperator) bool {
+	return op == qbtypes.FilterOperatorContains || op == qbtypes.FilterOperatorNotContains
+}
+
+func isMembershipLike(op qbtypes.FilterOperator) bool {
+	return op == qbtypes.FilterOperatorLike || op == qbtypes.FilterOperatorILike || op == qbtypes.FilterOperatorNotLike
+}
+
+type PathType struct {
+	Path     string `ch:"path"`
+	Type     string `ch:"type"`
+	LastSeen uint64 `ch:"last_seen"`
+}
+
+type JSONQueryBuilder struct {
+	telemetryStore       telemetrystore.TelemetryStore
+	cache                sync.Map // map[string]*utils.ConcurrentSet[telemetrytypes.JSONDataType]
+	lastSeen             uint64
+	promotedPaths        sync.Map     // map[string]struct{} of promoted JSON paths
+	stringIndexedColumns atomic.Value // map[string]string of string indexed columns
+
+	logger *slog.Logger
+}
+
+func NewJSONQueryBuilder(ctx context.Context,
+	telemetryStore telemetrystore.TelemetryStore, logger *slog.Logger) (*JSONQueryBuilder, error) {
+	builder := &JSONQueryBuilder{telemetryStore: telemetryStore, cache: sync.Map{}, logger: logger}
```

Comment:
add limit to cache ? take a good number that should be there and then add it ? also make it configurable maybe